pkg/kubelet/cm/qos_container_manager_linux.go

```diff
@@ -303,81 +306,92 @@ func (m *qosContainerManagerImpl) setMemoryQoS(configs map[v1.PodQOSClass]*Cgrou
 }
 
 func (m *qosContainerManagerImpl) UpdateCgroups() error {
-	m.Lock()
-	defer m.Unlock()
-
-	qosConfigs := map[v1.PodQOSClass]*CgroupConfig{
-		v1.PodQOSGuaranteed: {
-			Name:               m.qosContainersInfo.Guaranteed,
-			ResourceParameters: &ResourceConfig{},
-		},
-		v1.PodQOSBurstable: {
-			Name:               m.qosContainersInfo.Burstable,
-			ResourceParameters: &ResourceConfig{},
-		},
-		v1.PodQOSBestEffort: {
-			Name:               m.qosContainersInfo.BestEffort,
-			ResourceParameters: &ResourceConfig{},
-		},
-	}
+	_, err, _ := m.group.Do("qos", func() (interface{}, error) {
+		m.Lock()
+		defer m.Unlock()
+
+		// All update requests made until this point will wait for the current call.
+		// activePods callback is called after Forget, this ensures that pods
+		// added/removed before UpdateCgroups call are taken into account
+		// New update requests after Forget will trigger a new function invocation
+		// that will wait until the mutex that we acquired above is released
+		m.group.Forget("qos")
```

Comment:
It's strange to call methods on a single flight group while inside of the do closure. Is there a way we can avoid doing that and not use mutex locks in combination with single flight?

Reply:
The initial implementation used a queue to batch multiple requests and I changed it to use idiomatic code. I don't know other solution other than implementing something similar to what happens in single flight group under the hood for this use case, but I wanted to avoid duplicating code if possible.